src/producer/index.js

```diff
@@ -16,13 +17,28 @@ module.exports = ({
   cluster,
   logger: rootLogger,
   createPartitioner = createDefaultPartitioner,
-  retry = { retries: 5 },
+  retry,
+  idempotent = false,
 }) => {
+  retry = retry || (idempotent ? { retries: Number.MAX_SAFE_INTEGER } : { retries: 5 })
+
+  if (idempotent && retry.retries < 1) {
+    throw new KafkaJSNonRetriableError(
+      'Idempotent producer must allow retries to protect against transient errors'
+    )
+  }
+
+  const logger = rootLogger.namespace('Producer')
+
+  if (idempotent && retry.retries < Number.MAX_SAFE_INTEGER) {
+    logger.warn('Limiting retries for the idempotent producer may invalidate EoS guarantees')
+  }
+
   const partitioner = createPartitioner()
   const retrier = createRetry(Object.assign({}, cluster.retry, retry))
   const instrumentationEmitter = new InstrumentationEventEmitter()
-  const logger = rootLogger.namespace('Producer')
-  const sendMessages = createSendMessages({ logger, cluster, partitioner })
+  const transactionManager = createTransactionManager({ logger, cluster })
+  const sendMessages = createSendMessages({ logger, cluster, partitioner, transactionManager })
```

Comment:
Currently I pass `transactionManager` to the factory method. I do wonder if ultimately it will make more sense to provide a transaction manager in the `sendMessages` call. Currently it doesn't really matter but it may be something to refactor in #173

Reply:
In what way would it make the transaction manager less stateful? Maybe I'm missing something, but passing it into the factory method seems much more appropriate than passing it with each sendMessage call.

Reply:
You're probably right @Nevon, it's just that once we introduce transactions the transaction manager will become a bit of a state machine and I was trying to think of strategies to minimize the state we have to manage. One way of minimizing that would to be to have the transaction manager fetch the producer id on instantiation, so that there's no ambiguity about whether the idempotent producer has already be initialized. Since that fetch occur after connection that would probably mean passing the transaction manager into the send messages method, or lazily creating the sendMessages instance. Ultimately I do think what I've implemented is the most straight forward solution, with the benefit that it closely mirrors how transactions are managed in other Kafka libraries. Thus far in practice I haven't had an issue working with the state machine ([see preview here](https://github.com/tulios/kafkajs/blob/eos-173-transactional-producer/src/producer/transactionManager.js), just thinking through the alternatives.